ddtrace/internal/import_hooks/patch.py

```diff
@@ -0,0 +1,224 @@
+"""
+Module for patching Python's internal import system for import hooks.
+
+There are a few ways to monitor Python imports, but we have found the follow method
+of patching internal Python functions and builtin functions the most reliable and
+least invassive process.
+
+Given the challenges faced by implementing PEP 302, and our experience with monkey patching
+monkey patching and wrapping internal Python functions was an easier and more reliable way
+of implementing this feature in a way that has little effect on other packages.
+
+For us to successfully use import hooks, we need to ensure that the hooks are always called
+regardless of the condition or means of importing the module.
+
+
+PEP 302 defines a process for adding import "hooks".
+https://www.python.org/dev/peps/pep-0302/
+
+The way it works is by adding a custom "finder" onto `sys.meta_path`,
+for example: `sys.meta_path.append(MyFinder)`
+
+Finders are a way for users to customize the way the Python import system
+finds and loads modules.
+
+This approach has a few flaws
+
+1) Finders are called in order. In order for us to always ensure we are able to capture
+every import we would need to make sure that our Finder is always first in `sys.meta_path`.
+
+This can become tricky when the user uses another package that adds to `sys.meta_path`
+(e.g. `six`, or `newrelic`).
+
+If another Finder runs before ours and finds the module they are looking for, then ours will
+never get called.
+
+2) Finders were never meant to be used this way. An example of a good finder is a custom
+finder that looks for and loads a module off of a shared central server instead of a
+location on the existing Python path, or one that knows how to load modules from a `.zip`
+or another file format.
+
+3) The code is a little complex. In order to write a no-op finder we need to ensure
+other finders are called by us, and then at the end we look if they found a module and return it.
+
+This is the approach `wrapt` went for and requires locking and keep state of the current module
+being loaded and re-calling `__import__` for the module to trigger the other finders.
+
+4) Reloading a module is a weird case that doesn't always trigger a module finder.
+
+
+For these reasons we have decided to patch Python's internal module loading functions instead.
+"""
+import sys
+
+import wrapt
+
+from ..logger import get_logger
+from .registry import hooks
+
+
+log = get_logger(__name__)
+
+
+def exec_and_call_hooks(module_name, wrapped, args, kwargs):
+    """
+    Helper used to execute the wrapped function with args/kwargs and then call any
+      module hooks for `module_name` after
+    """
+    try:
+        return wrapped(*args, **kwargs)
+    finally:
```

Comment:
Does it make sense to still try to invoke the hooks if the module fails to load? I see the possibility that we miiight sidetrack with someone trying to debug their app since a module will fail to load and then there will most likely be an accompanying ddtrace debug log message.

Reply:
I'll add a comment, `hooks.call()` will only actually calls hooks if the module was successfully loaded.

Should probably add a log entry saying "tried to call hooks for un-loaded module" or something.